### PR TITLE
[SPARK-33748][K8S] Respect environment variables and configurations for Python executables

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes:

- Respect `PYSPARK_PYTHON` and `PYSPARK_DRIVER_PYTHON` environment variables, or `spark.pyspark.python` and `spark.pyspark.driver.python` configurations in Kubernates just like other cluster types in Spark.

- Depreate `spark.kubernetes.pyspark.pythonVersion` and guide users to set the environment variables and configurations for Python executables.
    NOTE that `spark.kubernetes.pyspark.pythonVersion` is already a no-op configuration without this PR. Default is `3` and other values are disallowed.

- In order for Python executable settings to be consistently used, fix `spark.archives` option to unpack into the current working directory in the driver of Kubernates' cluster mode. This behaviour is identical with Yarn's cluster mode. By doing this, users can leverage Conda or virtuenenv in cluster mode as below:

   ```python
    conda create -y -n pyspark_conda_env -c conda-forge pyarrow pandas conda-pack
    conda activate pyspark_conda_env
    conda pack -f -o pyspark_conda_env.tar.gz
    PYSPARK_PYTHON=./environment/bin/python spark-submit --archives pyspark_conda_env.tar.gz#environment app.py
   ```

- Removed several unused or useless codes such as `extractS3Key` and `renameResourcesToLocalFS`

### Why are the changes needed?

- To provide a consistent support of PySpark by using `PYSPARK_PYTHON` and `PYSPARK_DRIVER_PYTHON` environment variables, or `spark.pyspark.python` and `spark.pyspark.driver.python` configurations.
- To provide Conda and virtualenv support via `spark.archives` options.

### Does this PR introduce _any_ user-facing change?

Yes:

- `spark.kubernetes.pyspark.pythonVersion` is deprecated.
- `PYSPARK_PYTHON` and `PYSPARK_DRIVER_PYTHON` environment variables, and `spark.pyspark.python` and `spark.pyspark.driver.python` configurations are respected.

### How was this patch tested?

Manually tested via:

```bash
minikube delete
minikube start --cpus 12 --memory 16384
kubectl create namespace spark-integration-test
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ServiceAccount
metadata:
  name: spark
  namespace: spark-integration-test
EOF
kubectl create clusterrolebinding spark-role --clusterrole=edit --serviceaccount=spark-integration-test:spark --namespace=spark-integration-test
dev/make-distribution.sh --pip --tgz -Pkubernetes
resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh --spark-tgz `pwd`/spark-3.2.0-SNAPSHOT-bin-3.2.0.tgz  --service-account spark --namespace spark-integration-test
```

Unittests were also added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun and @erikerlandson FYI

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -836,21 +856,6 @@ private[spark] class SparkSubmit extends Logging {
     (childArgs.toSeq, childClasspath.toSeq, sparkConf, childMainClass)
   }
 
-  private def renameResourcesToLocalFS(resources: String, localResources: String): String = {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:839]: I read the codes multiple times for sure, and I think this code does a duplicated job.
If I am not horribly wrong somewhere:
- `localResources` itself are always local files, and `resources` will always be replaced to `localResources`.
-  If `resources` is `null, then `localResources` will be `null` too

cc @skonto from https://github.com/apache/spark/commit/5e74570c8f5e7dfc1ca1c53c177827c5cea57bf1

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh [hunk 2]
```diff
@@ -44,11 +44,11 @@ if [ -n "$SPARK_EXTRA_CLASSPATH" ]; then
   SPARK_CLASSPATH="$SPARK_CLASSPATH:$SPARK_EXTRA_CLASSPATH"
 fi
 
-if [ "$PYSPARK_MAJOR_PYTHON_VERSION" == "3" ]; then
-    pyv3="$(python3 -V 2>&1)"
-    export PYTHON_VERSION="${pyv3:7}"
```
Comment (HyukjinKwon) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:49]: `PYTHON_VERSION` and `pyv3` apparently not used anywhere in Spark code base.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala [hunk 3]
```diff
@@ -175,40 +175,81 @@ private[spark] trait DepsTestsSuite { k8sSuite: KubernetesSuite =>
...
+        expectedDriverLogOnCompletion = expectedDriverLogs,
         appArgs = Array("python3"),
         driverPodChecker = doBasicDriverPyPodCheck,
         executorPodChecker = doBasicExecutorPyPodCheck,
         appLocator = appLocator,
         isJVM = false,
-        pyFiles = Option(depsFile)) })
-  }
-
-  private def extractS3Key(data: String, key: String): String = {
```
Comment (HyukjinKwon) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:204]: `extractS3Key` is not used too.

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -387,20 +387,40 @@ private[spark] class SparkSubmit extends Logging {
...
+          val resolvedUris = Utils.stringToSeq(uris).map(Utils.resolveURI)
+          val localArchives = downloadFileList(
+            resolvedUris.map(
+              UriBuilder.fromUri(_).fragment(null).build().toString).mkString(","),
             targetDir, sparkConf, hadoopConf, secMgr)
-          UriBuilder.fromUri(downloadedUri).fragment(resolvedUri.getFragment).build().toString
+
+          // SPARK-33748: this mimics the behaviour of Yarn cluster mode. If the driver is running
+          // in cluster mode, the archives should be available in the driver's current working
+          // directory too.
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:403]: `spark.files` looks having the same issue too. However, I would like to avoid dealing all together in this PR.

File: core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala [hunk 5]
```diff
@@ -44,7 +44,7 @@ object PythonRunner {
       .orElse(sparkConf.get(PYSPARK_PYTHON))
       .orElse(sys.env.get("PYSPARK_DRIVER_PYTHON"))
       .orElse(sys.env.get("PYSPARK_PYTHON"))
-      .getOrElse("python")
+      .getOrElse("python3")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala:47]: This is rather a followup of https://github.com/apache/spark/commit/4f79b9fffd592a2f24e5e759d726232134bda409

**Comment by erikerlandson** [comment]
We should not remove the setting `spark.kubernetes.pyspark.pythonVersion`, as that will break backward compatibility. IMO, this setting should override environment variables, if it is specified.

**Comment by HyukjinKwon** [comment]
@erikerlandson, `spark.kubernetes.pyspark.pythonVersion` has no other values able to set except `3`, which is the default. Even if we remove `spark.kubernetes.pyspark.pythonVersion`, Spark still allows to set the non-existent configuration. Therefore I believe there's no breaking change here. In addition, Kubernates is experimental so far which generally accepts breaking changes between minor releases.

File: core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala:47]: Then, shall we proceed as another PR for `SPARK-32447 Use python3 by default in pyspark and find-spark-home scripts` because this PR seems to be under discussion.

**Review by dongjoon-hyun — COMMENTED** [review]
@erikerlandson and @HyukjinKwon .
This PR seems to contain 3 othorognal stuffs.
1. First of all, can we discuss of explicit deprecating `spark.kubernetes.pyspark.pythonVersion` as no-op at Apache Spark 3.1? Since it's effectively no-op for now and @HyukjinKwon 's new approach may supersede it, I believe it's okay.
2. `python` -> `python3` as new follow-up PR for SPARK-32447.
3. Finally, @HyukjinKwon 's proposal.

File: launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java [hunk 6]
```diff
@@ -336,7 +336,7 @@ private void checkJavaOptions(String javaOptions) {
       conf.get(SparkLauncher.PYSPARK_PYTHON),
       System.getenv("PYSPARK_DRIVER_PYTHON"),
       System.getenv("PYSPARK_PYTHON"),
-      "python"));
+      "python3"));
```
Comment (dongjoon-hyun) [inline launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java:339]: ditto. (As a new follow-up PR for SPARK-32447)

File: python/pyspark/context.py [hunk 7]
```diff
@@ -227,7 +227,7 @@ def _do_init(self, master, appName, sparkHome, pyFiles, environment, batchSize,
         os.environ["SPARK_BUFFER_SIZE"] = \
             str(self._jvm.PythonUtils.getSparkBufferSize(self._jsc))
 
-        self.pythonExec = os.environ.get("PYSPARK_PYTHON", 'python')
+        self.pythonExec = os.environ.get("PYSPARK_PYTHON", 'python3')
```
Comment (dongjoon-hyun) [inline python/pyspark/context.py:230]: ditto. (As a new follow-up PR for SPARK-32447)

File: core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala (hunk 5 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala:47]: Sure, makes sense. I'll create a separate PR tomorrow.

**Comment by HyukjinKwon** [comment]
Yes @erikerlandson thanks for clarification @dongjoon-hyun. So, if users set the configuration as 2 in 2.4 Spark application, that already wouldn't work because Python 2 was dropped. If users set it as 3, it will not be broken and work as expected because the default Python is already 3. Additionally now Spark 3.1 respects the Python environment variables if set, which is supposed to be, like Spark does in other places. I believe its more a bug fix.

File: docs/running-on-kubernetes.md [hunk 8]
```diff
@@ -1083,14 +1083,6 @@ See the [configuration page](configuration.html) for information on Spark config
...
   <td>2.4.0</td>
 </tr>
-<tr>
-  <td><code>spark.kubernetes.pyspark.pythonVersion</code></td>
-  <td><code>"3"</code></td>
-  <td>
-   This sets the major Python version of the docker image used to run the driver and executor containers. Can be 3.
-  </td>
-  <td>2.4.0</td>
-</tr>
```
Comment (viirya) [inline docs/running-on-kubernetes.md:1093]: The disappearance of the `pythonVersion` might make end users a bit hard to follow quickly. Seems no migration guide is there for kubernetes. Should we mention the removing and the support of PYSPARK_PYTHON and PYSPARK_DRIVER_PYTHON env variables in this doc?

Comment (erikerlandson) [reply docs/running-on-kubernetes.md:1093]: `pythonVersion` should (imo) go through a standard deprecation cycle, even if it's technically a no-op. That should include documentation and `@deprecated`

**Comment by erikerlandson** [comment]
What should be the precedence of these environment vars relative to `spark.pyspark.python` and `spark.pyspark.driver.python`?  The standard [precedence rules](https://spark.apache.org/docs/latest/configuration.html#dynamically-loading-spark-properties) don't actually cover alternate environment-variable forms. My take would be that the env-vars should have lower precedence than `spark.pyspark.python`, etc.

**Comment by HyukjinKwon** [comment]
@erikerlandson, these environment variables are documented at https://spark.apache.org/docs/latest/configuration.html#environment-variables.

Okay, let me just deprecate this with a warning message for now. We can remove it in the next release. That works to me.

File: docs/running-on-kubernetes.md (hunk 8 above)
Comment (HyukjinKwon) [reply docs/running-on-kubernetes.md:1093]: @erikerlandson, just to extra clarify, there's no need to `@deprecated`. The configuration instance is not an API. It's `private[spark]`. It would have been nicer to clarify what's the actual gain of deprecating over removing a no-op configuration.

**Comment by HyukjinKwon** [comment]
I believe this is a known flaky test (SPARK-33276), and it's hardly related:

```
- Test basic decommissioning with shuffle cleanup *** FAILED ***
  The code passed to eventually never returned normally. Attempted 182 times over 3.0065354390166665 minutes. Last failure message: "++ id -u
```

and all other tests passed. 

I believe I addressed all standing commands and it's ready for a review or possibly ready to go.

**Comment by HyukjinKwon** [comment]
cc @tgravescs too FYI

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 4 above)
Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:402]: did you file a separate lira then?

**Comment by tgravescs** [comment]
overall approach looks fine to me, I didn't have time to do detailed review.
I'm agree with deprecated config even if not used so users see warning.

I see we changed the driver env settings here, how is executor pyspark being set? I took a very quick look and didn't see it in BasicExecutorFeatureStep. Just wondering if that needs to be updated at all. Especially if driver python path was different from executor

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 4 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:402]: Sure, will file a JIRA.

**Comment by HyukjinKwon** [comment]
Thanks @tgravescs. If guiding users is a matter, we can still do by updating migration guide or even show a warning that the configuration was removed.

Given that this is experimental, I wanted to treat a bit differently as the removal and behaviour changes are expected as noted in the tags and as discussed in the mailing list in the past.

I feel like we're treating GA-ed feature and experimental feature similarly.

In any event, I agree that this is more conservative and possibly smoother approach. I'm okay with deprecating.

For setting envioronment variables kn executors, j believe PYSPARK_DRIVER_PYTHON is for driver python and PYSPARK_PYTHON is for executor python (and driver if PYSPARK_DRIVER_PYTHON is not set). So, if differenr python executables are used, they could set both environment variables (or equivalent Spark configurations such as `spark.pyspark.python`).

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 4 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:402]: SPARK-33782

**Comment by tgravescs** [comment]
sorry if I missed it in the PR or existing code, I want to make sure the executor side of spark.pyspark.python or PYSPARK_PYTHON is working with k8s?   I only saw changes in this PR for the driver side but haven't had time to look in great detail.

**Comment by HyukjinKwon** [comment]
Oh, actually `PYSPARK_PYTHON` (or `spark.pyspark.python`)  is being passed though from driver to executor side via:

1. https://github.com/apache/spark/blob/e2cdfcebd9b39a1104b34d8eafafbcdc6acf5d3e/python/pyspark/context.py#L230
2. https://github.com/apache/spark/blob/3959f0d9879fa7fa9e8f2e8ed8c8b12003d21788/python/pyspark/sql/udf.py#L35
3. https://github.com/apache/spark/blob/485145326a9c97ede260b0e267ee116f182cfd56/core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala#L64-L65
4. https://github.com/apache/spark/blob/3959f0d9879fa7fa9e8f2e8ed8c8b12003d21788/core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala#L92
5. https://github.com/apache/spark/blob/3959f0d9879fa7fa9e8f2e8ed8c8b12003d21788/core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala#L144

I also added a test to make sure this works in K8S here: https://github.com/apache/spark/pull/30735/files#diff-78ba045f393bcf6ffaa3dfe85bc7682cacf0bef69d447a2346e201279cc0bc5bR179-R197

**Comment by tgravescs** [comment]
great thanks!

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala [hunk 9]
```diff
@@ -153,22 +153,27 @@ object Utils extends Logging {
...
           new GzipCompressorOutputStream(
-            new FileOutputStream(
-              new File(outFile))))
+            new FileOutputStream(oFile)))
       ) { tOut =>
         val tarEntry = new TarArchiveEntry(fileToTarGz, fileToTarGz.getName)
+        // Each entry does not keep the file permission from the input file.
+        // Setting permissions in the input file do not work. Just simply set
+        // to 777.
+        tarEntry.setMode(0x81ff)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:170]: This looks like another orthogonal issue to me. Could you spin-off this one?

File: resource-managers/kubernetes/integration-tests/tests/py_container_checks.py [hunk 10]
```diff
@@ -24,7 +24,7 @@ def version_check(python_env, major_python_version):
         These are various tests to test the Python container image.
         This file will be distributed via --py-files in the e2e tests.
     """
-    env_version = os.environ.get('PYSPARK_PYTHON')
+    env_version = os.environ.get('PYSPARK_PYTHON', 'python3')
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/tests/py_container_checks.py:27]: This looks safe, but when this default value is used? I guess `PYSPARK_PYTHON` exists always. Is it just a fail-safe or an existing bug?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (two comments: https://github.com/apache/spark/pull/30735#discussion_r542888976 and https://github.com/apache/spark/pull/30735#discussion_r542889918).
I'll leave the decision for those comments to @HyukjinKwon .

**Comment by erikerlandson** [comment]
FWIW I also agree about separating the things not specifically about `PYSPARK_PYTHON`.
Otherwise LGTM

**Comment by dongjoon-hyun** [comment]
@erikerlandson and @viirya . I want to know your opinions. Please let us know if you have concerns still.

**Comment by dongjoon-hyun** [comment]
Oh, thanks, @erikerlandson !

File: resource-managers/kubernetes/integration-tests/tests/py_container_checks.py (hunk 10 above)
Comment (HyukjinKwon) [reply resource-managers/kubernetes/integration-tests/tests/py_container_checks.py:27]: Oh, when Spark application is submitted via `spark-submit`, looks its possible its unset. Without this change, the test fails.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala (hunk 9 above)
Comment (HyukjinKwon) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/Utils.scala:170]: Its slightly orthogonal but related to the current PR. The custom python in the tests should have executable permission and tgz should keep the permission. Without this change, the test fails because it loses executable permission and cannot be used as a python alone later. FWIW this util is for test-only, and this PR adds the first case in the test where it needs to keep the file permissions.

**Comment by HyukjinKwon** [comment]
Thanks @erikerlandson and @dongjoon-hyun for your review and approval. I replied in the comments.

**Comment by HyukjinKwon** [comment]
To clarify, I believe the current change contains all related changes. Without them, either test fails or python related env or configurations would not work as expected like other places.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1 (for K8S GA preparation).

Thanks again @erikerlandson, @dongjoon-hyun and @tgravescs for reviewing and and bearing with me :-).

**Comment by erikerlandson** [comment]
@HyukjinKwon thanks for standardizing the env-var support!

**Comment by HyukjinKwon** [comment]
Thank you @erikerlandson.
